orttraining/orttraining/python/training/ortmodule/_training_manager.py

```diff
@@ -228,7 +228,9 @@ def forward(self, *inputs, **kwargs):
             build_gradient_graph = False
             if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_BUILD_GRADIENT) is False or \
                     not self._onnx_models.exported_model:
+                random_states = _utils.get_random_states()
```

Comment:
Do we know why `self._export_model(...)` changes the random state? Is this introduced by the exporter or some code inside `ortmodule`?

Reply:
Probably because the models is executed once on that seed during export. After model is executed, the random number generator will run as it would for the second time

That is why we do a deep copy of the model before exporting, to make sure any randomness inside the model is not changed during export. 

However, I am not sure whether making ORTModule aware of external random genrators is a good idea. There can be many out there, including custom ones that ORTModule will never handle

It is probably a better idea to document such limitation that states that during first step random generators can be modified and it is user responsibility to reset them accordingly.

A sample extending `class ORTModulePreserveState(ORTModule)` could be provided to exemplify the issue and a possible solution

Reply:
@baijumeswani Yes [for this model](https://github.com/microsoft/onnxruntime/pull/10705#issue-1154987620) we found that the random state is changed during the shape inference: it is introduced by `sympy.Min()` within [`_infer_Slice`](https://github.com/microsoft/onnxruntime/blob/50a6f095cd2a13e524fba054cf82cfc76fa57185/onnxruntime/python/tools/symbolic_shape_infer.py#L1425).

Reply:
just for my understanding, are the different ranks performing different operations? If all the ranks are doing the same work, they should all have the same state after export. Maybe I am missing something here.

Reply:
@baijumeswani Different ranks should be performing the same operations. But the operation `sympy.Min()` is found to randomly change the random state, e.g. for the following simple case, starting from the same random state, the random state is different after `Min()` anyway. This explains why it results in different random states on different ranks. The root cause is inside those sympy functions, I think it is safer to do the save&restore on ORT side.

```py
import random
from sympy import Min, Symbol

a, b = Symbol('a'), Symbol('b')

random.seed(0)
state_before_0 = random.getstate()
Min(a, b)
state_after_0 = random.getstate()

random.seed(0)
state_before_1 = random.getstate()
Min(a, b)
state_after_1 = random.getstate()

print(state_before_0 == state_before_1) # True
print(state_after_0 == state_after_1) # False
```

Reply:
ok, thanks for the explanation.